### PR TITLE
HHH-15868 Expose more standard HQL functions through HibernateCriteriaBuilder

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-15868

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java [hunk 1]
```diff
@@ -2569,6 +2571,498 @@ private static SessionFactory locateSessionFactoryOnDeserialization(String uuid,
...
+				sqmArguments,
+				getTypeConfiguration().getBasicTypeForJavaType( type ),
+				queryEngine,
+				getJpaMetamodel().getTypeConfiguration()
+		);
+	}
+
+	@Override
+	public SqmFunction<String> format(Expression<? extends TemporalAccessor> datetime, String pattern) {
+		SqmExpression<String> patternLiteral = value( pattern );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java:2591]: I actually believe that this must be a literal:

```suggestion
		SqmExpression<String> patternLiteral = literal( pattern );
```

**Review by beikov — COMMENTED** [review]
Great work!

**Comment by beikov** [comment]
Look into `FunctionTests` which shows how some tests need to be skipped on certain databases. You need to add some `@RequiresDialectFeature` annotations.

File: hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmCriteriaNodeBuilder.java]: Actually it had to be a `SqmFormat` to allow for dialect translation of the pattern, now it's fixed.

**Comment by mbladel** [comment]
Thank you @beikov, I added dialect checks to tests and made some other small adjustments.

File: hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java [hunk 2]
```diff
@@ -356,6 +357,11 @@ public void initializeFunctionRegistry(QueryEngine queryEngine) {
 				new DB2FormatEmulation( queryEngine.getTypeConfiguration() )
 		);
 
+		queryEngine.getSqmFunctionRegistry().register(
+				"atan2",
+				new DB2Atan2Function( queryEngine.getTypeConfiguration() )
+		);
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:363]: While testing I noticed that DB2 requires the parameters to the `atan2` function in reverse order. This change fixes it, don't know if there's a better/more standard way of doing this.

Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java:363]: Usually we use a pattern for this purpose:

```
queryEngine.getSqmFunctionRegistry().patternDescriptorBuilder("atan2", "atan2(?2,?1)")
	.setInvariantType( queryEngine.getTypeConfiguration().getBasicTypeRegistry().resolve( StandardBasicTypes.DOUBLE ) )
	.setExactArgumentCount( 2 )
	.setParameterTypes(NUMERIC, NUMERIC )
	.register();
```

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/dialect/DB2Dialect.java]: Thanks for the tip, I thought there was a better way to do it. Updated it now.

**Review by beikov — APPROVED** [review]

**Pull request merged**